File: 0179-largest-number/0179-largest-number.cpp
#include <vector>
#include <string>
#include <algorithm>

class Solution {
public:
    std::string largestNumber(std::vector<int>& nums) {
        std::vector<std::string> num_strs;
        num_strs.reserve(nums.size());
        for (int num : nums) {
            num_strs.push_back(std::to_string(num));
        }
        
        std::sort(num_strs.begin(), num_strs.end(), [](const std::string& a, const std::string& b) {
            return a + b > b + a;
        });
        
        if (num_strs[0] == "0") {
            return "0";
        }
        
        std::string largest_num = "";
        for (const std::string& str : num_strs) {
            largest_num += str;
        }
        
        return largest_num;
    }
};


// Synced seamlessly with LeetHub Pro
// Pro features: https://bit.ly/leethubpro | Free version: https://bit.ly/leethubv4
// Get it here: https://chromewebstore.google.com/detail/bcilpkkbokcopmabingnndookdogmbna

File: 0179-largest-number/0179-largest-number.py
from functools import cmp_to_key

class Solution:
    def largestNumber(self, nums: list[int]) -> str:
        num_strs = [str(num) for num in nums]
        
        def compare(x: str, y: str) -> int:

            if x + y > y + x:
                return -1
            elif x + y < y + x:
                return 1
            else:
                return 0
        
        num_strs.sort(key=cmp_to_key(compare))
        
        largest_num = "".join(num_strs)
        
        return "0" if largest_num[0] == "0" else largest_num


# Synced seamlessly with LeetHub Pro
# Pro features: https://bit.ly/leethubpro | Free version: https://bit.ly/leethubv4
# Get it here: https://chromewebstore.google.com/detail/bcilpkkbokcopmabingnndookdogmbna